### PR TITLE
Add Shoebill Finance

### PR DESCRIPTION
**Pull request by ShoebillFinance**
**NOTE**

### Please enable "Allow edits by maintainers" while putting up the PR.
____

## Type of change
- [x] New adapters
- [ ] Adapter-related bug fix

Description:

___

### Description

Please provide the following information based on the type of points calculation:

#### For TVL Points Calculation:

1. Dapp Description:
   

  Lending Protocol (fork of Compound V2)

3. Criteria for Calculating User Points:
   

Users will earn points according to user's sbToken contract balance.

4. Collateral Asset Locking Contract:
   
   > **Note**: If there are multiple `tokenAddress`, please specify which contracts the different `tokenAddress` are locked in.

sb Token Addresses (holding collateral underlying assets)

ETH : 0x4dA697a89ea1D166881362b56E6863294820eC97
wstETH : 0xF52bB5C5F728a29c508f050dbC26e79eCfD82981
STONE.manta : 0x2060A2f15507e9722b5561516532399c5cAB475c
rsETH : 0x62873A9A5eec2379150d4309053532cb20767E4d
pufETH : 0x59d9CeAf0108A5af76336cAD0ec7C7C5c266Eb39
mswETH : 0x3b00F3f046f895415894d1a43e99628813E2816C
mmETH : 0x433102DC1694B37641C4ED0bE9B05a0A31495013
ezETH.linea : 0x3bf00732082c6Bcc92db19520B44dcC8a34ef586
ezETH.eth : 0xD661C805b98ecA6b214286D4d9AcF9D303845725
ezETH.arb : 0xa4f05Ff63017EC2B75972223a534f705f3586225

5. LP Contract:
   
   > **Note**: User balance is calculated as LP percentage multiplied by total collateral.

### CODE REVIEW

**Comment by xsteadybcgo** [comment]
image
The data with a balance of 0 is not expected. There should not be data where userAddress and poolAddress are the same

File: src/adapters/shoebill/subgraph/subgraph.yaml [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+schema:
+  file: ./schema.graphql
+dataSources:
+  - kind: ethereum
+    name: LayerBank
+    network: zklink-nova
+    source:
+      abi: LayerBankCore
+      address: '0x89e348e0ab4D83aEDbf1354004552822a805a12e'
+      startBlock: 2335536
```
Comment (xsteadybcgo) [inline src/adapters/shoebill/subgraph/subgraph.yaml:11]: Set the startBlock to 2335136 as it was created at this [block](https://explorer.zklink.io/tx/0x04f6609ef2c36a1b98ffc27c0e4a34932bcb8b21fc73b27a054d0b2f1e397740)

File: src/adapters/shoebill/subgraph/subgraph.yaml [hunk 2]
```diff
@@ -0,0 +1,57 @@
+specVersion: 1.0.0
+schema:
+  file: ./schema.graphql
+dataSources:
+  - kind: ethereum
+    name: LayerBank
+    network: zklink-nova
+    source:
+      abi: LayerBankCore
```
Comment (xsteadybcgo) [inline src/adapters/shoebill/subgraph/subgraph.yaml:9]: Standardize the naming by changing LayerBankCore to a shoebill-related term

File: src/adapters/shoebill/execution/src/sdk/fetch.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  let data;
+  let errors;
+  let retry = true;
+  let retryCount = 0;
+  const maxRetries = 10;
+
+  while (retry && retryCount < maxRetries) {
+    try {
+      const response = await fetch("http://3.114.68.110:8000/subgraphs/name/shoebill-points", {
+        method: "POST",
```
Comment (xsteadybcgo) [inline src/adapters/shoebill/execution/src/sdk/fetch.ts:15]: set is as a variable from `.env` file.

File: src/adapters/shoebill/subgraph/src/index.ts [hunk 4]
```diff
@@ -0,0 +1,93 @@
+import { MarketListed } from '../generated/LayerBank/LayerBankCore';
+import { LayerBankLToken, Transfer } from '../generated/templates/LayerBankLToken/LayerBankLToken';
+import { PoolTokenPosition, Pool } from '../generated/schema';
+import { LayerBankLToken as LayerBankLTokenTemplate } from '../generated/templates';
+import { Address, BigInt, Bytes, log } from '@graphprotocol/graph-ts';
+import { setUserInvalid, updateUserBalance } from './general';
+import { fetchTokenSymbol } from './utils/tokenHelper';
+import { ADDRESS_ZERO } from './utils/constants';
+
+const CEther = '0x4da697a89ea1d166881362b56e6863294820ec97';
```
Comment (xsteadybcgo) [inline src/adapters/shoebill/subgraph/src/index.ts:10]: Delete this unused variable

**Comment by xsteadybcgo** [comment]
Regarding the data, I checked the position of the pool address 0xf52bb5c5f728a29c508f050dbc26e79ecfd82981. The total amount of wstETH.Ethereum in the pool does not match the sum of all users' amounts in the CSV file, they should be equal

image
image

File: src/adapters/shoebill/execution/src/sdk/fetch.ts (hunk 3 above)
Comment (ShoebillFinance) [reply src/adapters/shoebill/execution/src/sdk/fetch.ts:19]: fixed thx;

**Comment by ShoebillFinance** [comment]
> Regarding the data, I checked the position of the pool address 0xf52bb5c5f728a29c508f050dbc26e79ecfd82981. The total amount of wstETH.Ethereum in the pool does not match the sum of all users' amounts in the CSV file, they should be equal
> 
> image image

image

I just fixed issue with tracking real underlying balances of liquidity pool

**Comment by xsteadybcgo** [comment]
> > Regarding the data, I checked the position of the pool address 0xf52bb5c5f728a29c508f050dbc26e79ecfd82981. The total amount of wstETH.Ethereum in the pool does not match the sum of all users' amounts in the CSV file, they should be equal
> > image image
> 
> image
> 
> I just fixed issue with tracking real underlying balances of liquidity pool

These data entries where the userAddress is also the poolAddress need to be filtered out

**Comment by ShoebillFinance** [comment]
> @ShoebillFinance These data entries where the userAddress is also the poolAddress need to be filtered out

Added filter in https://github.com/zkLinkProtocol/nova-point-v2/pull/66/commits/0861c014ecaba3e8088279d603b9e19885617a34

**Review by xsteadybcgo — APPROVED** [review]

**Pull request merged**